File: Sources/SettingsManager.swift
import Foundation

struct Replacement {
    var find: String
    var replace: String
}

class SettingsManager {
    static let shared = SettingsManager()

    private let settingsPath: String
    private var _replacements: [Replacement] = []

    var replacements: [Replacement] {
        get { _replacements }
        set {
            _replacements = newValue
            save()
        }
    }

    private init() {
        settingsPath = NSString("~/.custom-wispr-settings.json").expandingTildeInPath
        load()
    }

    func applyReplacements(_ text: String) -> String {
        var result = text
        for r in _replacements {
            guard !r.find.isEmpty else { continue }
            result = result.replacingOccurrences(of: r.find, with: r.replace, options: .caseInsensitive)
        }
        return result
    }

    private func load() {
        guard FileManager.default.fileExists(atPath: settingsPath),
              let data = FileManager.default.contents(atPath: settingsPath),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let arr = json["replacements"] as? [[String: String]] else {
            return
        }
        _replacements = arr.compactMap { dict in
            guard let find = dict["find"], let replace = dict["replace"] else { return nil }
            return Replacement(find: find, replace: replace)
        }
    }

    private func save() {
        let arr = _replacements.map { ["find": $0.find, "replace": $0.replace] }
        let json: [String: Any] = ["replacements": arr]
        guard let data = try? JSONSerialization.data(withJSONObject: json, options: .prettyPrinted) else {
            log("ERROR: Failed to serialize settings")
            return
        }
        FileManager.default.createFile(atPath: settingsPath, contents: data, attributes: [.posixPermissions: 0o600])
    }
}

File: windows/CustomWispr/Core/SettingsManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Windows.Forms;

namespace CustomWispr.Core;

internal class SettingsManager
{
    public static readonly SettingsManager Shared = new();

    private static readonly string SettingsPath = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
        ".custom-wispr-settings.json");

    public List<Replacement> Replacements { get; private set; } = new();
    public bool HasCompletedOnboarding { get; set; }
    public bool LaunchAtLogin { get; set; }

    // Hotkey settings — default: Ctrl+Win
    public Keys HotkeyKey { get; set; } = Keys.LWin;
    public Keys HotkeyModifiers { get; set; } = Keys.Control;

    private SettingsManager()
    {
        Load();
    }

    public string ApplyReplacements(string text)
    {
        var result = text;
        foreach (var r in Replacements)
        {
            if (string.IsNullOrEmpty(r.Find)) continue;
            result = result.Replace(r.Find, r.Replace ?? "", StringComparison.OrdinalIgnoreCase);
        }
        return result;
    }

    public void Load()
    {
        if (!File.Exists(SettingsPath)) return;

        try
        {
            var json = File.ReadAllText(SettingsPath);
            var data = JsonSerializer.Deserialize<SettingsData>(json);
            if (data != null)
            {
                Replacements = data.Replacements ?? new();
                HasCompletedOnboarding = data.HasCompletedOnboarding;
                LaunchAtLogin = data.LaunchAtLogin;
                if (!string.IsNullOrEmpty(data.HotkeyKey) && Enum.TryParse<Keys>(data.HotkeyKey, out var key))
                    HotkeyKey = key;
                if (!string.IsNullOrEmpty(data.HotkeyModifiers) && Enum.TryParse<Keys>(data.HotkeyModifiers, out var mods))
                    HotkeyModifiers = mods;
            }
        }
        catch (Exception ex)
        {
            Logger.Log($"Failed to load settings: {ex.Message}");
        }
    }

    public void Save()
    {
        try
        {
            var data = new SettingsData
            {
                Replacements = Replacements,
                HasCompletedOnboarding = HasCompletedOnboarding,
                LaunchAtLogin = LaunchAtLogin,
                HotkeyKey = HotkeyKey.ToString(),
                HotkeyModifiers = HotkeyModifiers.ToString()
            };

            var json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(SettingsPath, json);
        }
        catch (Exception ex)
        {
            Logger.Log($"Failed to save settings: {ex.Message}");
        }
    }

    public class Replacement
    {
        public string Find { get; set; } = "";
        public string Replace { get; set; } = "";
    }

    private class SettingsData
    {
        public List<Replacement>? Replacements { get; set; }
        public bool HasCompletedOnboarding { get; set; }
        public bool LaunchAtLogin { get; set; }
        public string? HotkeyKey { get; set; }
        public string? HotkeyModifiers { get; set; }
    }
}
